contractcourt/channel_arbitrator.go

```diff
@@ -1298,6 +1307,118 @@ func (c *ChannelArbitrator) isPreimageAvailable(hash lntypes.Hash) (bool,
+	// to see if the set difference of our HTLCs is non-empty. If so, then
+	// we may need to cancel back some HTLCs if we decide go to chain.
+	remoteDanglingActions := c.checkRemoteDanglingActions(
+		height, activeHTLCs, commitsConfirmed,
+	)
+
+	// Finally, we'll merge the two set of chain actions.
+	localActions := make(ChainActionMap)
+	for chainAction, htlcs := range localCommitActions {
+		localActions[chainAction] = append(
+			localActions[chainAction], htlcs...,
+		)
+	}
+	for chainAction, htlcs := range remoteDanglingActions {
+		localActions[chainAction] = append(
+			localActions[chainAction], htlcs...,
+		)
+	}
+
+	return localActions, nil
+}
+
+// checkRemoteChainActions examines the set of remote commitments for any HTLCs
+// that are close to timing out. If we find any, then we'll return a set of
+// chain actions for HTLCs that are on our commitment, but not theirs to cancel
+// immediately.
+func (c *ChannelArbitrator) checkRemoteDanglingActions(
+	height uint32, activeHTLCs map[HtlcSetKey]htlcSet,
+	commitsConfirmed bool) ChainActionMap {
+
+	var (
+		pendingRemoteHTLCs []channeldb.HTLC
+		localHTLCs         = make(map[uint64]struct{})
+		remoteHTLCs        = make(map[uint64]channeldb.HTLC)
+		actionMap          = make(ChainActionMap)
+	)
+
+	// First, we'll construct two sets of the outgoing HTLCs: those on our
+	// local commitment, and those that are on the remote commitment(s).
+	for htlcSetKey, htlcs := range activeHTLCs {
+		if htlcSetKey.IsRemote {
+			for _, htlc := range htlcs.outgoingHTLCs {
+				remoteHTLCs[htlc.HtlcIndex] = htlc
+			}
+		} else {
+			for _, htlc := range htlcs.outgoingHTLCs {
+				localHTLCs[htlc.HtlcIndex] = struct{}{}
+			}
+		}
+	}
+
+	// With both sets constructed, we'll now compute the set difference of
+	// our two sets of HTLCs. This'll give us the HTLCs that exist on the
+	// remote commitment transaction, but not on ours.
+	for htlcIndex, htlc := range remoteHTLCs {
+		if _, ok := localHTLCs[htlcIndex]; ok {
+			continue
+		}
+
+		pendingRemoteHTLCs = append(pendingRemoteHTLCs, htlc)
+	}
+
+	// Finally, we'll examine all the pending remote HTLCs for those that
+	// have expired. If we find any, then we'll recommend that they be
+	// failed now so we can free up the incoming HTLC.
+	for _, htlc := range pendingRemoteHTLCs {
+		// We'll now check if we need to go to chain in order to cancel
+		// the incoming HTLC.
+		goToChain := c.shouldGoOnChain(
+			htlc.RefundTimeout, c.cfg.OutgoingBroadcastDelta,
+			height,
+		)
+
+		// If we don't need to go to chain, and no commitments have
+		// been confirmed, then we can move on. Otherwise, if
+		// commitments have been confirmed, then we need to cancel back
+		// *all* of the pending remote HTLCS.
+		if !goToChain && !commitsConfirmed {
+			continue
+		}
+
+		log.Tracef("ChannelArbitrator(%v): immediately "+
+			"htlc=%x about to timeout on remote commitment",
+			c.cfg.ChanPoint, htlc.RHash[:])
```

Comment:
Log text reads not as a sentence.

Reply:
I don't think we have such a hard requirement. If so many other log messages would need to be modified. This is an existing log message with a swap swapped in.

Reply:
Ok, not always clear to see in the diff what was pre-existing.